examples/step-facet-laplace/step-12.cc

```diff
@@ -0,0 +1,703 @@
+
+                // +nu penalty g v
+                + nu[point] * penalty * g[point] *
+                    fe_face.shape_value(i, point)) *
+              JxW[point];
+        }
+    };
+
+    auto face_worker = [&](const Iterator &    cell,
+                           const unsigned int &f,
+                           const unsigned int &sf,
+                           const Iterator &    ncell,
+                           const unsigned int &nf,
+                           const unsigned int &nsf,
+                           ScratchData<dim> &  scratch_data,
+                           CopyData &          copy_data) {
+      FEFacetValues<dim> &fe_facet = scratch_data.fe_facet_values;
+      fe_facet.reinit(cell, f, sf, ncell, nf, nsf);
+      const auto &q_points = fe_facet.get_fe_values().get_quadrature_points();
+
+      copy_data.face_data.emplace_back();
+      CopyDataFace &copy_data_face = copy_data.face_data.back();
+
+      const unsigned int n_dofs        = fe_facet.n_facet_dofs();
+      copy_data_face.joint_dof_indices = fe_facet.get_facet_dof_indices();
+
+      copy_data_face.cell_matrix.reinit(n_dofs, n_dofs);
+
+      const std::vector<double> &        JxW = fe_facet.get_JxW_values();
+      const std::vector<Tensor<1, dim>> &normals =
+        fe_facet.get_normal_vectors();
+
+      std::vector<double> nu(q_points.size());
+      viscosity_function.value_list(q_points, nu);
+
+      const double degree =
+        std::max(1.0,
+                 static_cast<double>(fe_facet.get_fe_values().get_fe().degree));
+      const double extent1 =
+        cell->extent_in_direction(GeometryInfo<dim>::unit_normal_direction[f]) *
+        (cell->has_children() ? 2.0 : 1.0);
+      const double extent2 = ncell->extent_in_direction(
+                               GeometryInfo<dim>::unit_normal_direction[nf]) *
+                             (ncell->has_children() ? 2.0 : 1.0);
+      const double penalty =
+        4.0 * degree * (degree + 1.0) * 0.5 * (1.0 / extent1 + 1.0 / extent2);
+
+      for (unsigned int point = 0; point < q_points.size(); ++point)
+        {
+          for (unsigned int i = 0; i < n_dofs; ++i)
+            for (unsigned int j = 0; j < n_dofs; ++j)
+              copy_data_face.cell_matrix(i, j) +=
+                (
+                  // - nu {\nabla u}.n [v] (consistency)
+                  -nu[point] *
+                    (fe_facet.scalar().gradient_average(j, point) *
+                     normals[point]) *
+                    fe_facet.scalar().jump(i, point)
+
+                  // - nu [u] {\nabla v}.n  (symmetry) // NIPG: use +
+                  - nu[point] * fe_facet.scalar().jump(j, point) *
+                      (fe_facet.scalar().gradient_average(i, point) *
+                       normals[point])
+
+                  // nu sigma [u] [v] (penalty)
+                  + nu[point] * penalty * fe_facet.scalar().jump(j, point) *
+                      fe_facet.scalar().jump(i, point)
+
```

Comment:
I love how simple this looks like. It is how you write it down on paper, and even if we pay a little in terms of execution time, this is how I'd love to test DG things.

I'm not sure about the `.scalar()` notation. I'd just remove it, and make it conforming to our other fe_values  objects, where it is the user's responsability to know what component this is referring to, and if they really want to have something more complex, they should use extractors.